header-validator/src/validate-json.ts

```diff
@@ -118,289 +131,424 @@ type ListOpts = {
 }
 
 function list(
-  f: ValueCheck,
+  ctx: Context,
+  j: Json,
   { minLength = 0, maxLength = Infinity }: ListOpts = {}
-): ValueCheck {
-  return (ctx, values) => {
-    if (Array.isArray(values)) {
-      if (values.length > maxLength || values.length < minLength) {
-        ctx.error(`length must be in the range [${minLength}, ${maxLength}]`)
-      }
-
-      values.forEach((value, index) => ctx.scope(index, () => f(ctx, value)))
-      return
-    }
+): Json[] | undefined {
+  if (!Array.isArray(j)) {
     ctx.error('must be a list')
+    return
   }
+
+  if (j.length > maxLength || j.length < minLength) {
+    ctx.error(`length must be in the range [${minLength}, ${maxLength}]`)
+  }
+
+  return j
 }
 
-type Uint64Check = (ctx: Context, value: bigint) => void
+function uint64(ctx: Context, j: Json): bigint | undefined {
+  const s = string(ctx, j)
+  if (s === undefined) {
+    return
+  }
 
-function uint64(f: Uint64Check = () => {}): ValueCheck {
-  return string((ctx, value) => {
-    if (!uint64Regex.test(value)) {
-      ctx.error(`must be a uint64 (must match ${uint64Regex})`)
-      return
-    }
+  if (!uint64Regex.test(s)) {
+    ctx.error(`must be a uint64 (must match ${uint64Regex})`)
+    return
+  }
 
-    const int = BigInt(value)
-    const max = 2n ** 64n - 1n
-    if (int > max) {
-      ctx.error('must fit in an unsigned 64-bit integer')
-      return
-    }
+  const int = BigInt(s)
+  const max = 2n ** 64n - 1n
+  if (int > max) {
+    ctx.error('must fit in an unsigned 64-bit integer')
+    return
+  }
 
-    f(ctx, int)
-  })
+  return int
 }
 
-const triggerData = uint64((ctx, value) => {
-  if (typeof ctx.vsv !== 'undefined') {
-    Object.entries(ctx.vsv.triggerDataCardinality).forEach(([t, c]) => {
-      if (value >= c) {
-        ctx.warning(
-          `will be sanitized to ${
-            value % c
-          } if trigger is attributed to ${t} source`
-        )
-      }
-    })
+function triggerData(ctx: Context, j: Json): bigint | undefined {
+  const n = uint64(ctx, j)
+  if (n === undefined || ctx.vsv === undefined) {
```

Comment:
It doesn't enforce that, as evidenced by the lack of changes needed in the tests. If `vsv` is`undefined`, the value is still required to be a uint64, but no value is returned due to the inability to perform sanitization without knowing the vendor-specific value. That said, I could return the unsanitized value in that case.

Reply:
> the value is still required to be a uint64
Wouldn't it return undefined regardless of its value as soon as vsv is undefined ? I think the test still pass because the "trigger_data" field is optional (undefined == not there == invalid).

Reply:
`undefined` doesn't mean invalid by itself here. Validity for the purposes of the tests and the interactive form is solely based on how many errors are emitted to the context.

Reply:
> undefined doesn't mean invalid by itself here

Wouldn't you agree that this is confusing ? Everywhere else undefined  = invalid, except here where it means no validation was performed.

> That said, I could return the unsanitized value in that case.

I think this would make it more consistent.

Reply:
Done.